source/assets/js/playground/autocomplete.ts

```diff
@@ -0,0 +1,241 @@
+import {
+  CompletionContext,
+  CompletionResult,
+  CompletionSource,
+} from '@codemirror/autocomplete';
+import {sassCompletionSource} from '@codemirror/lang-sass';
+import {syntaxTree} from '@codemirror/language';
+import {EditorState} from '@codemirror/state';
+import moduleMembers from './module-members';
+
+// validFor identifier, from @codemirror/lang-css
+const identifier = /^(\w[\w-]*|-\w[\w-]*|)$/;
+
+// Sass-specific at rules only. CSS at rules should be added to `@codemirror/lang-css`.
+const atRuleKeywords = [
+  'use',
+  'forward',
+  'import',
+  'mixin',
+  'include',
+  'function',
+  'extend',
+  'error',
+  'warn',
+  'debug',
+  'at-root',
+  'if',
+  'else',
+  'each',
+  'for',
+  'while',
+];
+
+const atRuleOptions = Object.freeze(
+  atRuleKeywords.map(keyword => ({
+    label: `@${keyword} `,
+    type: 'keyword',
+    validFor: identifier,
+  }))
+);
+
+// Completions for Sass at rules
+function atRuleCompletion(context: CompletionContext): CompletionResult | null {
+  const atRule = context.matchBefore(/@\w*/);
+  if (!atRule) return null;
+  if (atRule.from === atRule.to && !context.explicit) return null;
+  return {
+    from: atRule.from,
+    to: atRule.to,
+    options: atRuleOptions,
+    validFor: identifier,
+  };
+}
+
+type CompletionInfo = {
+  name: string;
+  description?: string;
+};
+
+type CompletionModule = CompletionInfo & {
+  functions?: CompletionInfo[];
+  variables?: CompletionInfo[];
+};
+
+const moduleNames = moduleMembers.map(mod => mod.name);
+
+const moduleDescriptions = {
+  color:
+    'generates new colors based on existing ones, making it easy to build color themes',
+  list: 'lets you access and modify values in lists',
+  map: 'makes it possible to look up the value associated with a key in a map, and much more',
+  math: 'provides functions that operate on numbers',
+  meta: 'exposes the details of Sass’s inner workings',
+  selector: 'provides access to Sass’s powerful selector engine',
+  string: 'makes it easy to combine, search, or split apart strings',
+};
```

Comment:
I'd move this into `generate-module-members.ts` and inject it into the generated module metadata. Then not only do you have a single chunk of data for each module, but you also only have one source of truth where you're manually listing all the known modules.